### PR TITLE
cluster-logging: Add LokiStack tokenized auth proposal

### PR DESCRIPTION
**Pull request by periklis**
Based upon #1339 the following enhancement proposes the required changes in operating Loki Operator in STS-enabled clusters incl. HCP/ROSA.

Pre-requisites in upstream:
- grafana/loki/pull/11060

cc @xperimental @cahartma @jcantrill @alanconway @JoaoBraveCoding @btaani

### CODE REVIEW

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+As previously accomplished by other Red Hat and third-party operators (e.g. OADP Operator, External DNS Operator, AWS EFS CSI Driver Operator) the following proposal seeks to use the CCO's CredentialsRequest API in addition to the OpenShift Console OLM capabilities to detect STS-enabled clusters and manage the whole operator lifecycle.
+
+
+### User Stories
+
+* As a LokiStack administrator I want to off-load STS configuration to CCO so that I do not need to manually manage STS credentials configuration on AWS and in turn a custom LokiStack S3 object storage secret.
+* As a LokiStack administrator I want to off-load Azure WIF configuration to CCO so that I do not need to manually manage WIF credentials configuration on Azure and in turn a custom LokiStack Azure object storage secret.
+* As a LokiStack administrator I want to off-load GCP WIF configuration to CCO so that I do not need to manually manage WIF credentials configuration on GCP and in turn a custom LokiStack GCS object storage secret.
+
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:58]: Are we supporting only user stories that use CCO?

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+  secretRef:
+    name: logging-loki-aws-creds
+    namespace: openshift-logging
+  serviceAccountNames:
+  - logging-loki
+```
+
+#### Azure WIF Support
+
+The Loki Operator will use the provided environment variables (`CLIENT_ID`, `TENANT_ID`, `SUBSCRIPTION_ID`) and create a `CredentialsRequest` resource in the CCO operator namespace per LokiStack resource. The values will be used in the `CredentialsRequest` fields `spec.providerSpec.{AzureClientID,AzureTenantID,AzureSubscriptionID}` respectively. The LokiStack administrator is required to provider an Azure Region via the Azure object storage secret.
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:166]: There is a discrepancy between the fields mentioned here and in the [design doc](https://github.com/grafana/loki/pull/11060/files#diff-27b5aa2c87cf76afa33759f0ba55f749b2a54e1d4b1b6c7bc69a2cf1cbfc1c24R69-R76). In the design doc we mention `AzureClientID,AzureTenantID` however CCO seems to care about some other fields `AzureSubscriptionID, AzureRegion`.

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    - -runtime-config.file=/etc/loki/config/runtime-config.yaml
+    - -config.expand-env=true
+    env:
+    - name: AWS_ROLE_ARN
+      value: arn:aws:iam::269733383066:role/openshift-logging-loki-operator-controller-manager
+    - name: AWS_WEB_IDENTITY_TOKEN_FILE
+      value: /var/run/secrets/openshift/serviceaccount/token
+```
+
+in addition the `LokiStack` ConfigMap `logging-loki-config` needs to provid an `s3_storage_config` as below:
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:238]:
```suggestion
in addition the `LokiStack` ConfigMap `logging-loki-config` needs to provide an `s3_storage_config` as below:
```

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md (hunk 1 above)
Comment (jcantrill) [reply enhancements/cluster-logging/loki-tokenized-auth-enablement.md:58]: > Are we supporting only user stories that use CCO?

This verbiage implies the need to use the CCO utility but, aside from this proposal, it should be possible to use any flow that provides the necessary token.  My understanding is CCO simplifies the setup which results in a secret with the needed information.

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+## Proposal
+
+The Loki Operator supports per Cloud Provider the following annotations in it's respectives ClusterServiceVersion per bundle (community-openshift, openshift):
+
+- For AWS STS: `features.operators.openshift.io/token-auth-aws: true`
+- For Azure WIF: `features.operators.openshift.io/token-auth-azure: true`
+- For GCP WIF: `features.operators.openshift.io/token-auth-gcp: true`
+
+These will enable to operator installation to require specific IAM resource identifiers (e.g. role, client id, tenant id, etc.) which will be populated to the subscription and in turn exposed in the operator container environment.
```
Comment (jcantrill) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:79]: Some subtle grammar changes, but I would shorten this since it is long:

 These enable the operator installation to require specific IAM resource identifiers (e.g. role, client id, tenant id, etc.).  The operator subscription is populated with the identifiers and exposed in the operator container environment.

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+### Workflow Description
+
+The workflow implemented by this proposal follows exact the flow presented in the [Tokenized Authentication Enabledment for Red Hat Operators on Cloud Providers](tokenized-auth-enablement-workflow) enhancement proposal.
```
Comment (jcantrill) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:83]: ... exactly follows..

you could also just say "follows"

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+For the LokiStack administrator using the OpenShift Console - Operator installation under the proposed system:
+
+1. user goes to console and starts operator install
+2. console detects token-based (STS) cluster, and token-based-supporting operator
+3. console prompts user to create roles+supply ARN-type string
```
Comment (jcantrill) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:89]: roles and supply

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+4. console creates subscription with ARN-type string embedded as a spec.config.env
+5. operator deployment is created with ARN-type string in env
+6. operator creates `CredentialsRequest` including the ARN-type string
+7. Cloud Credential Operator populates Secret based on `CredentialsRequest`
+8. operator loads Secret, makes cloud service requests
```
Comment (jcantrill) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:94]: Secret and makes

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+
+For the LokiStack admnistator using the OpenShift CLI - Operator installation under the proposed system:
```
Comment (jcantrill) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:96]: ..administrator..

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md (hunk 2 above)
Comment (jcantrill) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:166]: ...is required to provide an Azure...

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+
+TBD along 4.15.
+
+### API Extensions
+
+No further API Extensions needed.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Based on the reconciled `CredentialRequest` resource the Loki Operator will populated the credential secrets as following plus additional configuration required by Loki to operate with tokenized authentication.
```
Comment (jcantrill) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:200]: ...secrets as follows...

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md (hunk 1 above)
Comment (periklis) [reply enhancements/cluster-logging/loki-tokenized-auth-enablement.md:58]: IMHO we should limit our support vector here because an operator bundle with the `features.operators.openshift.io/token-auth-{aws,azure,gcp}` annotation will only work in companion with the CCO workflow on OpenShift. Although we support upstream a setup w/o CCO, I suggest to support only the CCO workflow for OCP subscriptions.

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md (hunk 2 above)
Comment (periklis) [reply enhancements/cluster-logging/loki-tokenized-auth-enablement.md]: That's correct and well known because my investigation in the Loki code base did not uncover any usage of the azure subscription id and/or region. I am leaving this open here for the research task in LOG-4545. WDYT?

**Comment by bentito** [comment]
The EP looks really good -- it follows the referenced workflow with CCO and it provides practical implementation details  such as YAML configurations and environment variables which should make it a valuable resource for practical implementation _for other operator authors_ following this path to support tokenized auth on all three major cloud providers.

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+- For GCP: TBD along 4.15.
+
+For the Loki Operator author team:
+- Add code to create a `CredentialRequest` per `LokiStack` resource.
+- Add IAM-idetifier-type fields to the `CredentialRequest` resources.
+- Add eventing to report status on a CR to indicate lacking STS credentials for fully operational deploy or update.
+
+#### AWS STS Support
+
+The Loki Operator will use the `ROLEARN` environment variable and create a `CredentialsRequest` resource in the CCO operator namespace per LokiStack resource. The `ROLEARN` value is placed in the `spec.providerSpec.stsIAMRoleARN` field. The `CredentialsRequest` resource references the main LokiStack serviceaccount name that is the same as the LokiStack resource name.
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:135]:
```suggestion
The Loki Operator will use the `ROLEARN` environment variable and create a `CredentialsRequest` resource in the CCO operator namespace per LokiStack resource. The `ROLEARN` value is placed in the `spec.providerSpec.stsIAMRoleARN` field. The `CredentialsRequest` resource references the main LokiStack service account name which is the same as the LokiStack resource name. The LokiStack administrator is required to provide an AWS Region via the AWS object storage secret.
```

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+ name: ...
+spec:
+  config:
+    env:
+    - name: CLIENT_ID
+      value: "<azure client id>"
+    - name: TENANT_ID
+      value: "<azure tenant id>"
+    - name: SUBSCRIPTION_ID
+      value: "<azure subscription id>""
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:123]:
```suggestion
      value: "<azure subscription id>"
```

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+
+TBD along 4.15.
+
+### API Extensions
+
+No further API Extensions needed.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Based on the reconciled `CredentialRequest` resource the Loki Operator will populated the credential secrets as follows plus additional configuration required by Loki to operate with tokenized authentication.
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:200]: Rephrased feel free to adopt it or re-write it once more
```suggestion
The Loki Operator will generate the Loki configuration needed to operate with tokenized authentication based on the credential secret created by the reconciliation of the `CredentialRequest` resource. This Loki configuration will change depending on the Cloud Provider as follows.
```

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+    role_arn = <role ARN>
+    web_identity_token_file = <path to mounted service account token such /var/run/secrets/openshift/serviceaccount/token>
+kind: Secret
+metadata:
+  name: logging-loki-aws-credentials
+  namespace: openshift-logging
+type: Opaque
+```
+
+each Loki container will gain two new evironment variables:
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:222]:
```suggestion
Each Loki container will gain two new environment variables:
```

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+    - -runtime-config.file=/etc/loki/config/runtime-config.yaml
+    - -config.expand-env=true
+    env:
+    - name: AWS_ROLE_ARN
+      value: arn:aws:iam::269733383066:role/openshift-logging-loki-operator-controller-manager
+    - name: AWS_WEB_IDENTITY_TOKEN_FILE
+      value: /var/run/secrets/openshift/serviceaccount/token
+```
+
+in addition the `LokiStack` ConfigMap `logging-loki-config` needs to provide an `s3_storage_config` as below:
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:238]:
```suggestion
In addition the `LokiStack` ConfigMap `logging-loki-config` needs to provide an `s3_storage_config` as below:
```

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+  azure_subscription_id: <subscription id>
+  azure_tenant_id: <tenant id>
+kind: Secret
+metadata:
+  name: logging-loki-aws-credentials
+  namespace: openshift-logging
+type: Opaque
+```
+
+each Loki container will gain the following evironment variables:
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:269]:
```suggestion
Each Loki container will gain the following environment variables:
```

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+      value: <tenant id>
+    - name: AZURE_REGION
+      value: <azure region>
+    - name: AZURE_SUBSCRIPTION_ID
+      value: <subscription id>
+    - name: AZURE_FEDERATED_TOKEN_FILE
+      value: /var/run/secrets/openshift/serviceaccount/token
+```
+
+__Note:__ The Azure Region value is populated from the azure object storage secret referenced by the `lokistack.spec.storage.secret.name`.
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:291]: I think we can remove this, this seems to be from one of the first iterations of this document, region seems to also be in the secret created by CCO so we can use that.

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 17]
```diff
@@ -0,0 +1,343 @@
...
+
+#### Failure Modes
+
+#### Support Procedures
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:337]:
```suggestion
- AWS STS Support https://github.com/grafana/loki/pull/11481
```

**Review by JoaoBraveCoding — COMMENTED** [review]
Just some minor fixes.

My only big question about this is the fact that we require customers to pre-provision IAM resources, which IIUC, render CCO completely redundant as the Loki Operator with the RoleARN and some extra config from the storage secret would be able to make STS/WIF workflow work by itself. 
I say this because from my POV if an operator creates a `CredentialRequest`, this resource has all the necessary information for creating the IAM resources we are currently asking to be pre-provisioned. So currently our proposal, to me, feels like we are supporting STS for STS clusters in Manual mode and the other Modes I think it would also work but creating the `CredentialRequest` would only make CCO create extra IAM resources that wouldn't be used.
Maybe I'm misunderstanding something but I wanted to bring it up

**Comment by periklis** [comment]
> Just some minor fixes.
> 
> My only big question about this is the fact that we require customers to pre-provision IAM resources, which IIUC, render CCO completely redundant as the Loki Operator with the RoleARN and some extra config from the storage secret would be able to make STS/WIF workflow work by itself. I say this because from my POV if an operator creates a `CredentialRequest`, this resource has all the necessary information for creating the IAM resources we are currently asking to be pre-provisioned. So currently our proposal, to me, feels like we are supporting STS for STS clusters in Manual mode and the other Modes I think it would also work but creating the `CredentialRequest` would only make CCO create extra IAM resources that wouldn't be used. Maybe I'm misunderstanding something but I wanted to bring it up

To frame this a little bit more, we are not implementing all CCO modes of operation except the [short-lived tokens](https://github.com/openshift/cloud-credential-operator#4-short-lived-tokens) mode. This is the only one supported on all big three providers for the platform's operators. In addition the team behind CCO adds enablements (OCP Console OLM UI + CredentialsRequest CR additions) which we will gradually make use upon availability (i.e. [AWS since 4.14](https://github.com/openshift/cloud-credential-operator/pull/542) and [Azure in 4.15](https://github.com/openshift/cloud-credential-operator/pull/587)).

In fact in `short-lived-tokens` mode CCO plays a passive role only assuming that a cloud admin (e.g. GCP project admin) is managing IAM resources external to OCP and any other layered product operator. The small but net benefit for Loki Operator is to minimize the required inputs for the Object Storage Secret, i.e. comparing the two modes below:

1. Loki Operator on a Kubernetes cluster (somehow STS enabled but w/o a player like CCO), requires an object storage secret like:
```yaml
data:
  bucketnames: loki-bucket
  roleARN: the-full-role-arn-as-created-manually-by-the-cloud-admin
```
3. Loki Operator on OpenShift STS cluster running on CCO manual a.k.a. short-lived-tokens mode, requires an object storage secret like:
```yaml
data:
  bucketnames: loki-bucket
```

The reason not asking for a `RoleARN` in the latter case is that the CCO enablement (e.g. for AWS) asks for an upfront created role on Operator installation. In fact the operator and it's operands share the same role which mimics the pattern "I grant you roles, that I only possess".

In summary one needs to zoom out a bit and look on the CCO enablement from the UI to the operator lifecycle and lastly on the exchange of a `CredentialsRequest` for a credentials `Secret`. In fact anything that is IAM-related stays on the hands of the cloud administrator. The Operator installation is exposed only to the minimal set of those. Leaving the operands not needing any of those. This minimizes human interaction with any cloud IAM resources. Simply put:

_"Your don't need to know where you get your credentials for Loki to access. If your Cloud Admin created a role for you, Loki-Operator and CCO will make things travel for you."_

**Comment by JoaoBraveCoding** [comment]
> To frame this a little bit more, we are not implementing all CCO modes of operation except the [short-lived tokens](https://github.com/openshift/cloud-credential-operator#4-short-lived-tokens) mode.

Yeah this is the context I was missing/forgetting thanks a lot for the explanation

**Comment by dhellmann** [comment]
#1555 is changing the enhancement template in a way that will cause the header check in the linter job to fail for existing PRs. If this PR is merged within the development period for 4.16 you may override the linter if the only failures are caused by issues with the headers (please make sure the markdown formatting is correct). If this PR is not merged before 4.16 development closes, please update the enhancement to conform to the new template.

**Review by xperimental — COMMENTED** [review]
Updated to reflect the current implementation state of the Loki Operator.

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+TBD along 4.16.
+
+### Risks and Mitigations
+
+### Drawbacks
+
+## Design Details
+
+We have one PoC item:
+How Loki Operator can be changed to work with the modified CCO for AWS STS: https://github.com/grafana/loki/compare/main...periklis:loki:aws-sts-auth-spike
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:318]: This link should be updated to https://github.com/grafana/loki/pull/11060

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+  - logging-loki-ruler
+```
+
+#### GCP WIF Support
+
+TBD along 4.16.
+
+### API Extensions
+
+No further API Extensions needed.
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:198]: We should mention `CredentialMode` that was added recently

**Review by JoaoBraveCoding — APPROVED** [review]
lgtm

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 20]
```diff
@@ -0,0 +1,407 @@
...
+    // CredentialModeToken represents the usage of short-lived tokens retrieved from a credential source.
+    // In this mode the static configuration does not contain credentials needed for the object storage.
+    // Instead, they are generated during runtime using a service, which allows for shorter-lived credentials and
+    // much more granular control. This authentication mode is not supported for all object storage types.
+    CredentialModeToken CredentialMode = "token"
+    // CredentialModeManaged represents the usage of short-lived tokens retrieved from a credential source.
+    // This mode is similar to CredentialModeToken,but instead of having a user-configured credential source,
+    // it is configured by the environment, for example the Cloud Credential Operator in OpenShift.
+    // This mode is only supported for certain object storage types in certain runtime environments.
+    CredentialModeManaged CredentialMode = "managed"
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:220]:
```suggestion
    // CredentialModeTokenCCO represents the usage of short-lived tokens retrieved from a credential source.
    // This mode is similar to CredentialModeToken, but instead of having a user-configured credential source,
    // it is configured by the environment and the operator relies on the Cloud Credential Operator to provide
    // a secret. This mode is only supported for certain object storage types in certain runtime environments.
    CredentialModeTokenCCO CredentialMode = "token-cco"
```

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 21]
```diff
@@ -0,0 +1,407 @@
...
+GCP WIF support will be supported only in the form of the upstream support as per CCO missing respectable bits.
+
+### API Extensions
+
+The present proposal introduces a new optional field in the `ObjectStorageSecretSpec` namely `CredentialsMode`. If the user does not provide a value the `CredentialsMode` is automatically detected either from the provided secret fields or the operator environment variables (latter applies only in OpenShift cluters.). The selected or detected `CredentialsMode` is populated in addition in the `status.storage.credentialMode` field.
+
+```go
+/ CredentialMode represents the type of authentication used for accessing the object storage.
+//
+// +kubebuilder:validation:Enum=static;token;managed
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:204]:
```suggestion
// +kubebuilder:validation:Enum=static;token;token-cco
```

File: enhancements/cluster-logging/loki-tokenized-auth-enablement.md [hunk 22]
```diff
@@ -0,0 +1,407 @@
...
+    // CredentialMode can be used to set the desired credential mode for authenticating with the object storage.
+    // If this is not set, then the operator tries to infer the credential mode from the provided secret and its
+    // own configuration.
+    //
+    // +optional
+    // +kubebuilder:validation:Optional
+    CredentialMode CredentialMode `json:"credentialMode,omitempty"`
+}
+```
+The purpose of the `CredentialMode` is to override the detected credentials type from object storage secrets or the operator environment variables. Latter is only supported on AWS-STS/Azure-WIF managed OpenShift clusters, where the operator is using CredetialMode `managed` by default. However, the user might want to use `static` to store logs for example to Minio on the same cluster instead to AWS S3.
```
Comment (JoaoBraveCoding) [inline enhancements/cluster-logging/loki-tokenized-auth-enablement.md:235]:
```suggestion
The purpose of the `CredentialMode` is to override the detected credentials type from object storage secrets or the operator environment variables. Latter is only supported on AWS-STS/Azure-WIF managed OpenShift clusters, where the operator is using CredetialMode `token-cco` by default. However, the user might want to use `static` to store logs for example to Minio on the same cluster instead to AWS S3.
```

**Review by JoaoBraveCoding — COMMENTED** [review]
after this from my side, it's lgtm

**Review by JoaoBraveCoding — COMMENTED** [review]
more small fixes I noticed when giving it a final review

**Review by JoaoBraveCoding — APPROVED** [review]
/lgtm
